### PR TITLE
POCA trainer

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

This PR adds the POCA trainer and associated tests. In addition it makes changes to the extrinsic reward provider to enable team-based rewards to work. 

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

PR for documentation - to be merged after this one #5056
[Explanation](https://docs.google.com/document/d/1TJHImYY1N3a1xh_YOEWoY2aO81CygHdfbLs3IgeoXGo/edit#) of some of the design choices:

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: config/ppo/PushBlockCollab.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+behaviors:
```
Comment (ervteng) [inline config/ppo/PushBlockCollab.yaml:1]: TODO: remove this file and move to Hunter's pushblock branch

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
+# # Unity ML-Agents Toolkit
+# ## ML-Agent Learning (PPO)
+# Contains an implementation of PPO as described in: https://arxiv.org/abs/1707.06347
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/coma/trainer.py:3]: remove

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+from mlagents.trainers.policy.torch_policy import TorchPolicy
+from mlagents.trainers.coma.optimizer_torch import TorchCOMAOptimizer
+from mlagents.trainers.trajectory import Trajectory
+from mlagents.trainers.behavior_id_utils import BehaviorIdentifiers
+from mlagents.trainers.settings import TrainerSettings, PPOSettings
+
+logger = get_logger(__name__)
+
+
+class COMATrainer(RLTrainer):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:25]: Presumably this name is temporary, until we are on the official one, right?

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    """The COMATrainer is an implementation of the COMA2 algorithm."""
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:26]: This will be a great spot for the arxiv link :D

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        behavior_name: str,
+        reward_buff_cap: int,
+        trainer_settings: TrainerSettings,
+        training: bool,
+        load: bool,
+        seed: int,
+        artifact_path: str,
+    ):
+        """
+        Responsible for collecting experiences and training PPO model.
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:39]:
```suggestion
        Responsible for collecting experiences and training COMA model.
```

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    def get_policy(self, name_behavior_id: str) -> Policy:
+        """
+        Gets policy from trainer associated with name_behavior_id
+        :param name_behavior_id: full identifier of policy
+        """
+
+        return self.policy
+
+
+def lambda_return(r, value_estimates, gamma=0.99, lambd=0.8, value_next=0.0):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:301]: @andrewcoh Do you have a link to the equation this is implementing?

File: ml-agents/mlagents/trainers/settings.py [hunk 7]
```diff
@@ -215,6 +215,12 @@ def structure(d: Mapping, t: type) -> Any:
         return d_final
 
 
+@attr.s(auto_attribs=True)
+class ExtrinsicSettings(RewardSignalSettings):
+    # For use with COMA2. Add groupmate rewards to the final extrinsic reward.
+    add_groupmate_rewards = False
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/settings.py:221]: Is this a new hyperparameter for the user ?
Looks like the value is hard coded anyway in the coma trainer, is it really necessary ?

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 8]
```diff
@@ -333,3 +338,93 @@ def soft_update(source: nn.Module, target: nn.Module, tau: float) -> None:
                     alpha=tau,
                     out=target_param.data,
                 )
+
+    @staticmethod
+    def create_residual_self_attention(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:343]: I do not know how I feel about these two methods. I think they would look better as a module given how they are used.

File: ml-agents/mlagents/trainers/coma/trainer.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+            )
+
+            baseline_estimate = agent_buffer_trajectory[
+                RewardSignalUtil.baseline_estimates_key(name)
+            ].get_batch()
+            v_estimates = agent_buffer_trajectory[
+                RewardSignalUtil.value_estimates_key(name)
+            ].get_batch()
+
+            lambd_returns = lambda_return(
```
[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/coma/trainer.py:145]: GAE is basically the lambda return over n step advantages. It's possible we could reuse the `discount_rewards` function from the PPO trainer but I'm not sure it's worth putting that function in a new file just for this purpose.

File: ml-agents/mlagents/trainers/coma/trainer.py (hunk 6 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/coma/trainer.py:301]: This is a pretty good discussion of TD(lambda) I think. https://amreis.github.io/ml/reinf-learn/2017/11/02/reinforcement-learning-eligibility-traces.html

TD(lambda) was also used in COMA https://github.com/oxwhirl/pymarl/blob/master/src/utils/rl_utils.py#L4

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 10]
```diff
@@ -156,6 +105,166 @@ def forward(
...
+    def __init__(
+        self,
+        observation_specs: List[ObservationSpec],
+        network_settings: NetworkSettings,
+        action_spec: ActionSpec,
+    ):
+        super().__init__()
+        self.normalize = network_settings.normalize
+        self.use_lstm = network_settings.memory is not None
+        # Scale network depending on num agents
```
[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:118]: It's an artifact. removed.

File: ml-agents/mlagents/trainers/settings.py (hunk 7 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/settings.py:221]: It is not a new hyperparameter and it is actually to avoid requiring users to learn about a new group reward signal. POCA will set this to True on initialization so that the ExtrinsicReward provider knows to collect teammate rewards. Otherwise it will be False. The only way for users to change it is to modify python.

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 11]
```diff
@@ -156,6 +105,166 @@ def forward(
         return encoding, memories
 
 
+class MultiInputNetworkBody(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py]: Renamed to MultiAgentNetworkBody and added docstring.

I think you're right but that we should leave this for a refactor later on.

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 8 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:343]: Maybe we can combine this with a change to NetworkBody to also satisfy the comment that MultiAgentNetworkBody should reuse NetworkBody. The right solution may be to rename NetworkBody to Encoders that has the functionality of these methods and move the linear layer to the Actor/Critic. Then the MultiAgentNetBody may be able to reuse the encoders.

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:343]: reference https://github.com/Unity-Technologies/ml-agents/pull/5005#discussion_r591813207

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/networks.py]: Why leave this for later ? It is not a refactor, it is a new feature. And I think we should be careful with adding new refactor work for later.

File: ml-agents/mlagents/trainers/settings.py (hunk 7 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/settings.py:221]: Then why is it a setting and not a constant in the code?

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py]: I'm going to play devil's advocate here, and say it's probably not going to be a good idea to try and reuse networkbody. Originally I tried to do that during the prototyping phase and there were too many special cases. The dataflow is just different. 

A networkbody goes `obs -> encoders -> RSA -> linear layers -> output`. 
The MultiInput one has to do `obs -> concat with other obs -> encoders x num_obs -> RSA x num_obs -> another RSA -> linear layers -> output.`

So really what we're re-using is the RSA, the encoders, and the linear layer, but in a different order, and these are all already separate modules. To re-use NetworkBody we'd have to insert the 2nd RSA inside the first NetworkBody.

File: ml-agents/mlagents/trainers/settings.py (hunk 7 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:221]: The alternative would be to have a public variable that is set later (ew) or additional constructor parameter to Extrinsic Reward Provider (and only Extrinsic Reward Provider). This would mess with the BaseRewardProvider abstraction. I don't really have a preference, but it is choosing between the lesser of several evils here.

Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:221]: Potential alternative: #5082

**Review by ervteng — APPROVED** [review]
LGTM, will wait for at least one more reviewer

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+
+        def critic_pass(
+            self,
+            obs: List[List[torch.Tensor]],
+            memories: Optional[torch.Tensor] = None,
+            sequence_length: int = 1,
+        ) -> Tuple[torch.Tensor, torch.Tensor]:
+            """
+            A centralized value function. It calls the forward pass of MultiAgentNetworkBody
+            with just the states of all agents.
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:105]: Explain the meaning of the parameters (especially for `baseline` since it is not an interface implementation.

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+                actions=[],
+                memories=memories,
+                sequence_length=sequence_length,
+            )
+            value_outputs, critic_mem_out = self.forward(
+                encoding, memories, sequence_length
+            )
+            return value_outputs, critic_mem_out
+
+        def forward(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:119]: Remove this method. It has no reason to be public.

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+            old_baseline_values[name] = ModelUtils.list_to_tensor(
+                batch[RewardSignalUtil.baseline_estimates_key(name)]
+            )
+
+        n_obs = len(self.policy.behavior_spec.observation_specs)
+        current_obs = ObsUtil.from_buffer(batch, n_obs)
+        # Convert to tensors
+        current_obs = [ModelUtils.list_to_tensor(obs) for obs in current_obs]
+        group_obs = GroupObsUtil.from_buffer(batch, n_obs)
+        group_obs = [
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:240]: Some of my comments got lost. Please review them in the conversation tab : https://github.com/Unity-Technologies/ml-agents/pull/5005#discussion_r591824890

File: ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py [hunk 15]
```diff
@@ -5,11 +5,42 @@
...
-        return np.array(mini_batch[BufferKey.ENVIRONMENT_REWARDS], dtype=np.float32)
+        indiv_rewards = np.array(
+            mini_batch[BufferKey.ENVIRONMENT_REWARDS], dtype=np.float32
+        )
+        total_rewards = indiv_rewards
+        if (
+            BufferKey.GROUPMATE_REWARDS in mini_batch
+            and BufferKey.GROUP_REWARD in mini_batch
+        ):
+            if self.add_groupmate_rewards:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: Invert these 2 ifs. No need to check the first one if there are no groumaterewards

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: These if conditions could be better:

if self.add_groupmate_rewards and BufferKey.GROUPMATE_REWARDS in mini_batch : do the groupmate reward
if BufferKey.GROUP_REWARD in mini_batch : Do the group reward

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 16]
```diff
@@ -156,6 +105,181 @@ def forward(
...
+                ] = 0.0  # Remoove NaNs fast
+                no_nan_obs.append(new_obs)
+            obs_with_no_nans.append(no_nan_obs)
+        return obs_with_no_nans
+
+    def forward(
+        self,
+        obs_only: List[List[torch.Tensor]],
+        obs: List[List[torch.Tensor]],
+        actions: List[AgentAction],
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:224]: I cannot help but think there is a better way. Something like 
obs_without_action: List[List[torch.Tensor]],
obs_action: List[Tuple[List[torch.Tensor], List[torch.Tensor]]]
I know the types are ugly, but it should be explicit that the length of obs and actions has to be equal

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 17]
```diff
@@ -333,3 +338,148 @@ def soft_update(source: nn.Module, target: nn.Module, tau: float) -> None:
...
+                    1,
+                    hidden_size,
+                    kernel_init=Initialization.Normal,
+                    kernel_gain=(0.125 / hidden_size) ** 0.5,
+                )
+            rsa = ResidualSelfAttention(hidden_size, entity_num_max)
+        return rsa, x_self_encoder
+
+    @staticmethod
+    def encode_observations(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:376]: Stand by my statement, make create_residual_self_attention a module with encode_observations its forward method

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/utils.py:376]: Call it ObservationEncoder

File: ml-agents/mlagents/trainers/settings.py [hunk 18]
```diff
@@ -600,9 +600,14 @@ def _team_change_default(self):
...
     PPO: str = "ppo"
     SAC: str = "sac"
+    POCA: str = "poca"
 
     def to_settings(self) -> type:
-        _mapping = {TrainerType.PPO: PPOSettings, TrainerType.SAC: SACSettings}
+        _mapping = {
+            TrainerType.PPO: PPOSettings,
+            TrainerType.SAC: SACSettings,
+            TrainerType.POCA: PPOSettings,
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/settings.py:609]: For clarity define POCA settings as 
```
POCASettings = PPOSettings
```
Will set us up better when the two settings diverge

**Review by vincentpierre — APPROVED** [review]
I approve once all comments have been resolved. 
Including those that were wrongfully marked as outdated by github.

File: ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py (hunk 15 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: Updated

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 16 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:224]: I'm actually not in favor of having nested tuples as parameters, it doesn't enforce length to be equal any more than two parameters, and requires constructing additional datastructures before use.

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:296]: https://github.com/Unity-Technologies/ml-agents/pull/5094/

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 17 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py]: courtesty of @ervteng https://github.com/Unity-Technologies/ml-agents/pull/5093/
